### ISSUE TITLE
FTP connection "leak"

### ISSUE THREAD

**Issue by mylemans** [duplicati/duplicati#1688]
I'm trying to backup to OVH' backup FTP server, but it has a 3 concurrent connection limit (per user).

I tried to enable all options that "should" make duplicati only use one concurrent connection, but nothing seems to help.

In addition, I pressed the "test" button a couple of times and after that I kept getting the error.

I checked "netstat -natp" and I noticed it had multiple "ESTABLISHED" entries.
##

**Comment by mylemans** [comment]
I'm giving it a look myself. So far I cannot reproduce the issue on my local (windows) PC. Maybe it is a mono bug?

I'm currently running it on Ubuntu 14.04 using the "supplied" mono setup, but I also have a separate up2date mono installation on there, I'll give that a try and report back.

**Comment by mylemans** [comment]
Issue remains the same when using mono 3.2.8 on Ubuntu 14.04.

I pressed "Test" 3 times. Then after that sometimes I got "connection ok", sometimes the "user limit exceeded" message. After a couple of tries it is now stuck at "user limit exceeded".

edit: After giving it a minute I could press the test button twice before getting the 3 user limit reached message.

**Comment by sfarrell5123** [comment]
I think duplicati is making a new ftp connection for each file upload/download request. So it's an insane  number of small ftp connections. It's only holding 1 open at a time though, but it's probably causing the FTP server to trip its user limit

It's also making FTP really slow
here is a log snippet

Wed Feb 15 10:14:54 2017 [pid 997952] CONNECT: Client "10.0.1.66"
Wed Feb 15 10:14:54 2017 [pid 997951] [notes] OK LOGIN: Client "10.0.1.66"
Wed Feb 15 10:14:55 2017 [pid 997953] [notes] OK UPLOAD: Client "10.0.1.66", "/archive/duplicati/duplicati-b52d8dc9640e34a948cd62b78fedd233f.dblock.7z", 52381556 bytes, 106855.94Kbyte/sec
Wed Feb 15 10:14:55 2017 [pid 997955] CONNECT: Client "10.0.1.66"
Wed Feb 15 10:14:55 2017 [pid 997954] [notes] OK LOGIN: Client "10.0.1.66"
Wed Feb 15 10:14:58 2017 [pid 997958] CONNECT: Client "10.0.1.66"
Wed Feb 15 10:14:58 2017 [pid 997957] [notes] OK LOGIN: Client "10.0.1.66"
Wed Feb 15 10:14:58 2017 [pid 997959] [notes] OK UPLOAD: Client "10.0.1.66", "/archive/duplicati/duplicati-i84812a52d27248ed9d33d3726575a80a.dindex.7z", 18136 bytes, 6830.29Kbyte/sec
Wed Feb 15 10:14:58 2017 [pid 997961] CONNECT: Client "10.0.1.66"
Wed Feb 15 10:14:58 2017 [pid 997960] [notes] OK LOGIN: Client "10.0.1.66"
Wed Feb 15 10:15:01 2017 [pid 997964] CONNECT: Client "10.0.1.66"
Wed Feb 15 10:15:01 2017 [pid 997963] [notes] OK LOGIN: Client "10.0.1.66"
Wed Feb 15 10:15:02 2017 [pid 997965] [notes] OK UPLOAD: Client "10.0.1.66", "/archive/duplicati/duplicati-b2413fd083bb1488697f338f5c2b18eca.dblock.7z", 52381555 bytes, 63055.22Kbyte/sec
Wed Feb 15 10:15:02 2017 [pid 998030] CONNECT: Client "10.0.1.66"
Wed Feb 15 10:15:02 2017 [pid 998029] [notes] OK LOGIN: Client "10.0.1.66"

**Comment by ts678** [comment]
A lot of time has passed, but here's an effort to see what still seems to be there.

> I checked "netstat -natp" and I noticed it had multiple "ESTABLISHED" entries.

Not reproducible. I do get `TIME_WAIT`, but they shouldn't be causing a problem.
There's even a proper `QUIT`, `Goodbye`, and connection disconnect in server log.
Testing with FileZilla Server on Windows through 127.0.0.1 from a local Duplicati.

With a 1 connection limit, I tested backup, compact, and restore. They all worked.
Also tested a backup that did lots of uploads. They went fine -- but one at a time.

FTP (Alternative) seems to need a 2 connection limit, as it keeps first open to end.
I'm not sure why there's a difference, but I think I saw SFTP doing the same thing.

FTP connection "leak" doesn't seem to be there now. Last call for any other notes.
